md/barostat/moves.py

```diff
@@ -0,0 +1,196 @@
+from jax import config, numpy as jnp
+from jax.ops import segment_sum
+
+config.update("jax_enable_x64", True)
+
+import numpy as onp
+
+from md.states import CoordsVelBox
+from md.barostat.utils import compute_box_volume, compute_box_center
+
+from md.moves import MonteCarloMove
+
+from typing import List, Iterable, Tuple
+
+
+def compute_centroid(group):
+    return jnp.mean(group, axis=0)
+
+
+def _scatter_inds_from_group_inds(group_inds):
+    """
+    given a list of arrays of ints, representing groups of particles,
+    construct a flat array, representing which group index to sort each particle into
+
+    [[0,1,2], [3,4,5]] --> [0, 0, 0, 1, 1, 1]
+    """
+    all_inds = onp.hstack(group_inds)
+    scatter_inds = onp.zeros(len(all_inds))
+
+    # assert group_inds not overlapping
+    assert (len(all_inds) == len(set(all_inds)))
+
+    for i, group in enumerate(group_inds):
+        for j in group:
+            scatter_inds[j] = i
+
+    return jnp.array(scatter_inds, dtype=int)
+
+
+class CentroidRescaler:
+    def __init__(self, group_inds, weights=None):
+        self.group_inds = group_inds
+        self.group_sizes = jnp.array(list(map(len, self.group_inds)))
+        assert (jnp.min(self.group_sizes) > 0)
+
+        self.scatter_inds = _scatter_inds_from_group_inds(group_inds)
+
+        if weights is not None:
+            raise (NotImplementedError)
+
+    def rescale(self, coords, center, scale=1.0):
+        """scale distances of coords to center"""
+
+        dx_initial = coords - center
+        dx_updated = scale * dx_initial
+        return center + dx_updated
+
+    def compute_centroids(self, coords):
+        """Returns an array containing the centroids of each group"""
+        return segment_sum(coords, self.scatter_inds) / jnp.expand_dims(self.group_sizes, axis=1)
+
+    def _slow_compute_centroids(self, coords):
+        """For testing / reference"""
+        return jnp.array([compute_centroid(coords[g]) for g in self.group_inds])
+
+    def displace_by_group(self, coords, displacements):
+        return coords + displacements[self.scatter_inds]
+
+    def scale_centroids(self, coords, center, scale):
+        """
+
+        Notes
+        -----
+        * Currently ignores particle mass -- the centroid of a group of
```

Comment:
Does the OpenMM centroid computation take mass into account? @badisa

Reply:
(PS if we decide to support this, `np.average` as opposed to `np.mean` allows one to pass in a list of weights.

Reply:
Based on my reading of the OpenMM, it does not take the masses into account. That said, the kernel logic that they have is  fairly opaque to me. 

https://github.com/openmm/openmm/blob/be19e0222ddf66f612016a3c1f687161a53c2396/platforms/common/src/CommonKernels.cpp#L7625